core/src/main/java/com/linecorp/armeria/common/metric/ServiceNaming.java

```diff
@@ -0,0 +1,150 @@
+                    continue;
+                }
+                final String name = delegate.getClass().getName();
+                return name.substring(name.lastIndexOf('.') + 1);
+            }
+        };
+    }
+
+    /**
+     * Returns the {@link ServiceNaming} that returns the full name of innermost class from the given service.
+     * It will be fully qualified name including the package name separated by a period.
+     * e.g. {@code com.foo.HelloService}.
+     *
+     * @see Class#getName()
+     */
+    static ServiceNaming fullTypeName() {
+        return ctx -> {
+            final RpcRequest rpcReq = ctx.rpcRequest();
+            if (rpcReq != null) {
+                final String serviceType = rpcReq.serviceType().getName();
+                if ("com.linecorp.armeria.internal.common.grpc.GrpcLogUtil".equals(serviceType)) {
+                    // Parse gRPC serviceName and methodName
+                    final String fullMethodName = rpcReq.method();
+                    final int methodIndex = fullMethodName.lastIndexOf('/');
+                    return fullMethodName.substring(0, methodIndex);
+                } else {
+                    return serviceType;
+                }
+            }
+
+            Unwrappable unwrappable = ctx.config().service();
+            for (;;) {
+                final Unwrappable delegate = unwrappable.unwrap();
+                if (delegate != unwrappable) {
+                    unwrappable = delegate;
+                    continue;
+                }
+                return delegate.getClass().getName();
+            }
+        };
+    }
+
+    /**
+     * Converts the specified {@linkplain RequestOnlyLog#serviceName() serviceName}
+     * into another service name which is used as a meter tag or distributed trace's span name.
+     *
+     * <p>A naming rule can be set by either {@link ServerBuilder#defaultServiceNaming(ServiceNaming)} or
+     * {@link ServiceBindingBuilder#defaultServiceNaming(ServiceNaming)}.
+     * One of pre-defined naming rules is able to be used as follows.
+     * <h2>Example</h2>
+     * <pre>{@code
+     * Server server = Server
+     *   .builder()
+     *   .service(...)
+     *   .defaultServiceNaming(ServiceNaming.simpleTypeName())
+     *   .build()
+     * }</pre>
+     *
+     * <h2>Example 2</h2>
+     * <pre>{@code
+     * Server server = Server
+     *   .builder()
+     *   .route().path("/")
+     *   .defaultServiceNaming(ServiceNaming.fullTypeName())
+     *   ...
+     *   .build()
+     * }</pre>
+     *
+     * <p>If customizing is needed out of given rules, a lambda expression would be applied as follows.
+     * <pre>{@code
+     * Server server = Server
+     *   .builder()
+     *   .service(...)
+     *   .defaultServiceNaming(ctx -> {
+     *     return ctx.config().server().defaultHostname() + ctx.config().route().patternString();
+     *   })
+     *   .build()
+     * }</pre>
+     */
+    String convert(ServiceRequestContext ctx);
```

Comment:
Is it better to allow returning a null value and if it is null, use the default service naming?